.github/workflows/lua-tests.yml

```diff
@@ -10,19 +10,90 @@ on:
       - name: Checkout repository
         uses: actions/checkout@v4
       
-      - name: Cache Lua dependencies
+      - name: Set up Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
+          cache: 'npm'
+      
+      - name: Install Node.js dependencies
+        run: npm ci
+      
+      - name: Log in to Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      
+      - name: Try to pull pre-built image
+        id: pull-image
+        run: |
+          echo "Attempting to pull pre-built mGBA image..."
+          if docker pull ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:latest 2>/dev/null; then
+            echo "✅ Successfully pulled pre-built image"
+            echo "image_available=true" >> $GITHUB_OUTPUT
+            echo "image_source=ghcr" >> $GITHUB_OUTPUT
+            # Tag it for local use
+            docker tag ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:latest mgba-test-env:latest
+          elif docker pull ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:main 2>/dev/null; then
+            echo "✅ Successfully pulled main branch image"
+            echo "image_available=true" >> $GITHUB_OUTPUT
+            echo "image_source=ghcr" >> $GITHUB_OUTPUT
+            # Tag it for local use
+            docker tag ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}:main mgba-test-env:latest
+          else
+            echo "⚠️ Pre-built image not available, will build from source"
+            echo "image_available=false" >> $GITHUB_OUTPUT
+            echo "image_source=build" >> $GITHUB_OUTPUT
+          fi
+        continue-on-error: true
+      
+      - name: Build mGBA image from source (fallback)
+        if: steps.pull-image.outputs.image_available != 'true'
+        run: |
+          echo "Building mGBA image from source..."
+          docker build -f docker/Dockerfile -t mgba-test-env:latest .
+          echo "✅ Built mGBA image from source"
+      
+      - name: Start mGBA container for testing
+        run: |
+          echo "Starting mGBA container..."
+          docker run -d --name mgba-test \
+            -p 7102:7102 \
+            -e DISPLAY=:99 \
+            -e QT_QPA_PLATFORM=xcb \
+            mgba-test-env:latest
+          
+          # Wait for container to be ready
+          echo "Waiting for mGBA HTTP server to be ready..."
+          timeout 60s bash -c 'until curl -sf http://localhost:7102/ > /dev/null 2>&1; do sleep 2; done'
+          echo "✅ mGBA HTTP server is ready"
+      
+      - name: Verify mGBA container health
+        run: |
+          echo "Testing mGBA HTTP server endpoints..."
+          curl -sf http://localhost:7102/ || { echo "❌ Root endpoint failed"; exit 1; }
+          echo "✅ mGBA HTTP server is responding correctly"
+      
+      - name: Cache Lua dependencies (backup)
         uses: actions/cache@v4
```

Comment:
if image is not available for download, we do not want to build the docker image here. we rather just use the lua mock environment as a fallback. we also might need to adjust the test files to support the lua mock environment aswell as the real mgba docker env